### PR TITLE
wayland: Cache scale so that first frame respects HiDPI

### PR DESCRIPTION
**Pull request by Nefsen402**
Requires: https://github.com/emersion/mako/pull/410

### CODE REVIEW

**Comment by emersion** [comment]
If this is based on #393, can you add Co-authored-by trailers where applicable to preserve authorship?

**Comment by Nefsen402** [comment]
The last commit has an added Co-authored-by tag, the commit that I took direct inspiration from changes in #393.

**Comment by emersion** [comment]
I'm not a fan of "wayland: Cache scale so that first frame respects HiDPI", because it doesn't work in the general case. I'd prefer to reach a consensus on https://gitlab.freedesktop.org/wayland/wayland/-/issues/133 and fix the issue in all cases instead.

File: wayland.c [hunk 1]
```diff
@@ -376,11 +376,16 @@ static void layer_surface_handle_configure(void *data,
 		uint32_t serial, uint32_t width, uint32_t height) {
 	struct mako_surface *msurface = data;
 
+	zwlr_layer_surface_v1_ack_configure(surface, serial);
```
Comment (emersion) [inline wayland.c:379]: `ack_configure` must always be followed by `commit`. If the old buffer is suitable, we can perform a commit without attaching a new buffer.

**Comment by emersion** [comment]
"mako: Convert signed to unsigned types for sizes" → we need to be a bit careful about this, because surprising things happen with unsigned types. Add/substract to a signed integer for instance and you can hit overflows.

File: wayland.c (hunk 1 above)
Comment (Nefsen402) [reply wayland.c:379]: I tried to add a wl_surface_commit.

Comment (emersion) [reply wayland.c:379]: Yup, sounds good!

**Comment by Nefsen402** [comment]
> "mako: Convert signed to unsigned types for sizes" → we need to be a bit careful about this, because surprising things happen with unsigned types. Add/substract to a signed integer for instance and you can hit overflows.

I needed to make these changes because otherwise the same-size commit commit would not compile as unsigned numbers cannot be compared to signed ones. I thought casting wasn't a good option, so I changed the types.

> I'm not a fan of "wayland: Cache scale so that first frame respects HiDPI", because it doesn't work in the general case. I'd prefer to reach a consensus on https://gitlab.freedesktop.org/wayland/wayland/-/issues/133 and fix the issue in all cases instead.

It's definitely not ideal, but this commit was able to get me up and running and removing the jank with current mako. Maybe we can put the caching into its own PR so that the same-size commit changes can make their way through? The other changes improve  https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3352.

**Comment by Nefsen402** [comment]
> Yeah, I'd like to merge the same-size commit changes.

I pulled that out into its own PR, so that source branch names would make some sense.

**Comment by emersion** [comment]
Closing because this is better addressed by listening to the `preferred_buffer_scale` event and letting the compositor dictate the scale, rather than relying on Sway's specific implementation of enter events.

**Pull request closed without merging**